Validate MD integration methods groups

melodyyzh opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description
HOOMD fails to provide an error message when users provide invalid groups to the MD integration method. 
<!-- Describe your changes in detail. -->

## Motivation and context
Code to validate the rigid body flags exist but is not called. Defined `validate_group` to call the method. 
<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #1466 

## How has this been tested?
Will add a test to verify the output 
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->


## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

melodyyzh committed:
added  function to call  method

b-butler commented on hoomd/md/methods/methods.py:
```diff
@@ -27,10 +27,20 @@ class Method(AutotunedObject):
+        Users can call `validate_group` to verify the groups after changing
+        them.
+        """
+        self._cpp.obj.validateGroup()
```
```suggestion
        self._cpp_obj.validateGroup()
```

b-butler commented on hoomd/md/methods/methods.py:
```diff
@@ -27,10 +27,20 @@ class Method(AutotunedObject):
         self._simulation.state.update_group_dof()
 
+    def validate_group(self):
+        """Verify groups.
+
+        Groups may change after attaching.
+        Users can call `validate_group` to verify the groups after changing
+        them.
+        """
```
Change to clarify that only checks that constituent particles are not integrated.

melodyyzh committed:
Update hoomd/md/methods/methods.py

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

melodyyzh committed:
called methods and added tests

melodyyzh committed:
Merge branch 'validate-group' of https://github.com/glotzerlab/hoomd-blue into validate-group

melodyyzh commented:
I added functions to call `validateGroup` method in `methods.py` and `integrate.py`, and added tests to `test_methods.py` and `test_integrate.py`. Many tests failed when running pytest, one common message was "hoomd.md._md.IntegratorTwoStep' object has no attribute 'validateGroup'". I would appreciate some suggestions!

joaander commented:
> I added functions to call `validateGroup` method in `methods.py` and `integrate.py`, and added tests to `test_methods.py` and `test_integrate.py`. Many tests failed when running pytest, one common message was "hoomd.md._md.IntegratorTwoStep' object has no attribute 'validateGroup'". I would appreciate some suggestions!

You added the method to `DynamicIntegrator` which mirrors the C++ `IntegratorTwoStep`. `validateGroup` is in the C++ class `IntegrationMethodTwoStep` which is mirrored by `Method` in Python. We have two options:
1) Add a `validateGroups` method to `IntegratorTwoStep` in C++. This method could also validate that the groups of the contained methods do not intersect.
2) ~~Move your added code to `Method` and validate each method one at a time.~~

b-butler commented:
> You added the method to DynamicIntegrator which mirrors the C++ IntegratorTwoStep. validateGroup is in the C++ class IntegrationMethodTwoStep which is mirrored by Method in Python. We have two options:

@joaander that is on me. I thought there were two problems: particles being oversubscribed to multiple methods and constituent particles. I assumed we could check both and that both methods existed in HOOMD, and told @melodyyzh to assume such.

@melodyyzh to fix this you should create a method `validateGroups` in `IntegratorTwoStep` that checks that that the groups do not overlap. I would use [std::set_union](https://en.cppreference.com/w/cpp/algorithm/set_union) for this (take consecutive unions and check that the output length is the sum of the input lengths).

joaander commented:
> > You added the method to DynamicIntegrator which mirrors the C++ IntegratorTwoStep. validateGroup is in the C++ class IntegrationMethodTwoStep which is mirrored by Method in Python. We have two options:
> 
> @joaander that is on me. I thought there were two problems: particles being oversubscribed to multiple methods and constituent particles. I assumed we could check both and that both methods existed in HOOMD, and told @melodyyzh to assume such.
> 
> @melodyyzh to fix this you should create a method `validateGroups` in `IntegratorTwoStep` that checks that that the groups do not overlap. I would use [std::set_union](https://en.cppreference.com/w/cpp/algorithm/set_union) for this (take consecutive unions and check that the output length is the sum of the input lengths).

No worries. It is a good idea to add the intersection test. v2 had it, but I don't see the code for it any more. There are group union and intersection methods you can use: https://github.com/glotzerlab/hoomd-blue/blob/d43142b264af58b87b5adf669640a7f878d95326/hoomd/ParticleGroup.cc#L672-L688 . 

`validateGroups` should also call `validateGroup` on each method.

joaander commented:
Correction: Use `groupUnion` to compute the unions as @b-butler suggested. This will be more efficient than using intersections: https://github.com/glotzerlab/hoomd-blue/blob/d43142b264af58b87b5adf669640a7f878d95326/hoomd/ParticleGroup.cc#L360-L367

joaander reviewed (commented):
We discussed the needed changes offline.

melodyyzh committed:
removed changes

melodyyzh committed:
create validateGroups method in IntegratorTwoStep

melodyyzh committed:
modified the validate_groups function in integrator.py

melodyyzh committed:
Merge branch 'trunk-patch' into validate-group

joaander commented on hoomd/md/integrate.py:
```diff
@@ -51,6 +51,7 @@ def _attach_hook(self):
         if self.rigid is not None:
             self.rigid._attach(self._simulation)
         super()._attach_hook()
+        self.validate_group()
```
```suggestion

    def _post_attach_hook(self):
        self.validate_groups()
```

Also, 1) add:
```
    def _post_attach_hook(self):
        """Hook called after applying parameter dicts."""
        pass
```
to _HOOMDBaseObject in `operation.py`, and 2) call `self._post_attach_hook()` at the end of `_HOOMDBaseObject.attach`

This set of changes will call `validate_groups` *after* the C++ `IntegratorTwoStep` instances is given the list of `methods`.

joaander commented on hoomd/md/pytest/test_integrate.py:
```diff
@@ -60,6 +60,41 @@ def test_detaching(make_simulation, integrator_elements):
     assert not integrator._constraints._synced
 
 
+def test_validate_group():
+    CUBE_VERTS = [
+        (-0.5, -0.5, -0.5),
+        (-0.5, -0.5, 0.5),
+        (-0.5, 0.5, -0.5),
+        (-0.5, 0.5, 0.5),
+        (0.5, -0.5, -0.5),
+        (0.5, -0.5, 0.5),
+        (0.5, 0.5, -0.5),
+        (0.5, 0.5, 0.5),
+    ]
+
+    rigid = hoomd.md.constrain.Rigid()
+    rigid.body['R'] = {
+        "constituent_types": ['A'] * 8,
+        "positions": CUBE_VERTS,
+        "orientations": [(1.0, 0.0, 0.0, 0.0)] * 8,
+    }
+
+    nve1 = hoomd.md.methods.ConstantVolume(filter=hoomd.filter.All())
+    nve2 = hoomd.md.methods.ConstantVolume(filter=hoomd.filter.All())
+    integrator = hoomd.md.Integrator(dt=0,
+                                     methods=[nve1, nve2],
+                                     integrate_rotational_dof=True)
+    integrator.rigid = rigid
+
+    sim = hoomd.Simulation(device=hoomd.device.CPU())
+    sim.create_state_from_gsd('init.gsd')
+
+    sim.operations.integrator = integrator
+
+    with pytest.raises(RuntimeError):
+        sim.run(10)
```
```suggestion
def test_validate_group(simulation_factory, two_particle_snapshot_factory):
    CUBE_VERTS = [
        (-0.5, -0.5, -0.5),
        (-0.5, -0.5, 0.5),
        (-0.5, 0.5, -0.5),
        (-0.5, 0.5, 0.5),
        (0.5, -0.5, -0.5),
        (0.5, -0.5, 0.5),
        (0.5, 0.5, -0.5),
        (0.5, 0.5, 0.5),
    ]

    rigid = hoomd.md.constrain.Rigid()
    rigid.body['R'] = {
        "constituent_types": ['A'] * 8,
        "positions": CUBE_VERTS,
        "orientations": [(1.0, 0.0, 0.0, 0.0)] * 8,
    }

    nve1 = hoomd.md.methods.ConstantVolume(filter=hoomd.filter.All())
    nve2 = hoomd.md.methods.ConstantVolume(filter=hoomd.filter.All())
    integrator = hoomd.md.Integrator(dt=0,
                                     methods=[nve1, nve2],
                                     integrate_rotational_dof=True)
    integrator.rigid = rigid

    snapshot = two_particle_snapshot_factory(particle_types=['R', 'A'])
    if snapshot.communicator.rank == 0:
        snapshot.particles.body[:] = [0, 1]
    
    sim = simulation_factory(snapshot)

    sim.operations.integrator = integrator

    with pytest.raises(RuntimeError):
        sim.run(10)
```
Generate the initial configuration via snapshot.

joaander commented on hoomd/md/pytest/test_integrate.py:
```diff
@@ -60,6 +60,41 @@ def test_detaching(make_simulation, integrator_elements):
+    with pytest.raises(RuntimeError):
+        sim.run(10)
+
+
```
Adjust the first test to use one `NVE` method on all particles (should fail because it includes constituent particles).

Add a 2nd test that uses two `NVE` methods on rigid center particles. This one should fail because it has overlapping groups.

joaander reviewed (commented):
Here are some additional changes needed once you get the C++ code to compile.

melodyyzh committed:
Merge branch 'trunk-patch' into validate-group

melodyyzh committed:
modified source code and test file

melodyyzh committed:
modified the test file

melodyyzh committed:
Update hoomd/md/integrate.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

melodyyzh committed:
modified source codes and test file

joaander commented on hoomd/md/IntegratorTwoStep.cc:
```diff
@@ -389,6 +389,31 @@ bool IntegratorTwoStep::areForcesAnisotropic()
 
+void IntegratorTwoStep::validateGroups()
+    {
+    // Check if user-provided groups overlap
+    size_t group_size = 0;
+    for (auto& method : m_methods)
+        {
+        method->validateGroup();
+        group_size += method->getGroup()->getNumMembersGlobal();
+        }
+    if (m_methods.size() <= 1)
+        {
+        return;
+        }
+    auto group_union
+        = ParticleGroup::groupUnion(m_methods[0]->getGroup(), m_methods[1]->getGroup());
```
```suggestion
    // Check that methods have valid groups.
    size_t group_size = 0;
    for (auto& method : m_methods)
        {
        method->validateGroup();
        group_size += method->getGroup()->getNumMembersGlobal();
        }

    // Check that methods have non-overlapping groups.   
    if (m_methods.size() <= 1)
        {
        return;
        }
    auto group_union
        = ParticleGroup::groupUnion(m_methods[0]->getGroup(), m_methods[1]->getGroup());
```

joaander reviewed (approved):
Thanks! I have one suggestion to make the C++ code comments more detailed.

melodyyzh committed:
Update hoomd/md/IntegratorTwoStep.cc

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

melodyyzh commented:
Just commit the changes. Thank you for all the help on this PR!

joaander commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

b-butler commented on hoomd/md/pytest/test_integrate.py:
```diff
@@ -60,6 +60,52 @@ def test_detaching(make_simulation, integrator_elements):
+                                     integrate_rotational_dof=True)
+    integrator.rigid = rigid
+
+    snapshot = two_particle_snapshot_factory(particle_types=['R', 'A'])
+    if snapshot.communicator.rank == 0:
+        snapshot.particles.body[:] = [0, 1]
+
+    sim = simulation_factory(snapshot)
+    rigid.create_bodies(sim.state)
+    sim.operations.integrator = integrator
+
+    with pytest.raises(RuntimeError):
+        sim.run(10)
+
+    nve1 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
+    nve2 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
+    integrator.methods = [nve1, nve2]
+
+    with pytest.raises(RuntimeError):
+        integrator.validate_groups()
```
Make this a second test since it is testing a different condition.

b-butler replied:
```suggestion

def test_overlapping_filters(simulation_factory, two_particles_snapshot_factory):
    snapshot = two_particle_snapshot_factory(particle_types=['R', 'A'])
    if snapshot.communicator.rank == 0:
        snapshot.particles.body[:] = [0, 1]

    sim = simulation_factory(snapshot)
    rigid.create_bodies(sim.state)
    sim.operations.integrator = integrator

    with pytest.raises(RuntimeError):
        sim.run(10)

    nve1 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
    nve2 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
    integrator.methods = [nve1, nve2]

    with pytest.raises(RuntimeError):
        integrator.validate_groups()
```

b-butler reviewed (commented):
Just one comment to address then this is ready to go.

melodyyzh committed:
Update hoomd/md/pytest/test_integrate.py

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

melodyyzh committed:
modified test file

melodyyzh committed:
modified test file

melodyyzh commented:
I fixed some typos and added `rigid` and `integrator` in the second test since they were not specified. The test fails because `test_overlapping_filters` did not raise a runtime error. Any suggestions on what I did not notice that I need to change?

melodyyzh committed:
modified the test

b-butler commented on hoomd/md/pytest/test_integrate.py:
```diff
@@ -86,21 +89,27 @@ def test_validate_groups(simulation_factory, two_particle_snapshot_factory):
     nve1 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
     nve2 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
     integrator.methods = [nve1, nve2]
+    sim.run(0)
```
Shouldn't this error as attaching will automatically run the validation?

b-butler replied:
Nevermind I see now you call run above. I would want to see that errors too though to ensure it errors on attaching.

joaander replied:
It *only* errors on attaching, not when setting new methods. This second call to `run` passes because the `Integrator` is already attached.

joaander replied:
I thought it worth testing this behavior specifically, I'll add a comment. The 2nd test verifies that `validate_groups` throws errors on attach.

b-butler commented on hoomd/md/pytest/test_integrate.py:
```diff
@@ -86,21 +89,27 @@ def test_validate_groups(simulation_factory, two_particle_snapshot_factory):
+    sim.run(0)
+
     nve1 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
     nve2 = hoomd.md.methods.NVE(filter=hoomd.filter.Tags([0, 1]))
```
```suggestion
    nve1 = hoomd.md.methods.NVE(filter=hoomd.filter.All())
    nve2 = hoomd.md.methods.NVE(filter=hoomd.filter.All())
```

joaander replied:
Both the current and suggested change trigger the appropriate exception.

b-butler reviewed (changes requested):
I am not sure why one of the tests does not fail since the groups should be checked at runtime and are overlapping.

joaander committed:
Explain tests.

joaander committed:
Use clear rigid constituent error message.

joaander committed:
Ensure that errors are raised on all ranks.

joaander committed:
Add back in missing create_bodies.

joaander committed:
Remove debug print.

melodyyzh commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
